src/gluonts/mx/trainer/callback.py

```diff
@@ -0,0 +1,623 @@
+
+        callback_types = set([type(callback) for callback in self.callbacks])
+        # make sure not to have no duplicates
+        for callback in new_callbacks:
+            if type(callback) in callback_types:
+                continue
+            else:
+                self.callbacks.append(callback)
+
+    def on_train_start(self, max_epochs: int) -> None:
+        for callback in self.callbacks:
+            callback.on_train_start(max_epochs=max_epochs)
+
+    def on_network_initializing_end(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        for callback in self.callbacks:
+            callback.on_network_initializing_end(
+                training_network=training_network
+            )
+
+    def on_train_epoch_start(self, training_network: nn.HybridBlock) -> None:
+        for callback in self.callbacks:
+            callback.on_train_epoch_start(training_network=training_network)
+
+    def on_validation_epoch_start(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        for callback in self.callbacks:
+            callback.on_validation_epoch_start(
+                training_network=training_network
+            )
+
+    def on_train_batch_end(self, training_network: nn.HybridBlock) -> None:
+        for callback in self.callbacks:
+            callback.on_train_batch_end(training_network=training_network)
+
+    def on_validation_batch_end(
+        self, training_network: nn.HybridBlock
+    ) -> None:
+        for callback in self.callbacks:
+            callback.on_validation_batch_end(training_network=training_network)
+
+    def on_train_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return np.all(
+            [
+                callback.on_train_epoch_end(
+                    epoch_no=epoch_no,
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_validation_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+    ) -> bool:
+        return np.all(
+            [
+                callback.on_validation_epoch_end(
+                    epoch_no=epoch_no,
+                    epoch_loss=epoch_loss,
+                    training_network=training_network,
+                    trainer=trainer,
+                )
+                for callback in self.callbacks
+            ]
+        )
+
+    def on_epoch_end(
+        self,
+        epoch_no: int,
+        epoch_loss: float,
+        training_network: nn.HybridBlock,
+        trainer: gluon.Trainer,
+        best_epoch_info: Dict[str, Any],
+        ctx: mx.Context,
+    ) -> bool:
+        return np.all(
```

Comment:
Looking at this, we could also do something like:

```py
    def _juxt(self, name, *args, **kwargs):
        fns = map(attrgetter(name), self.callbacks)
        return toolz.juxt(*fns)(*args, **kwargs)

    def on_epoch_end(
        self,
        *args, **kwargs
    ) -> bool:
        return all(self._juxt("on_epoch_end", *args, **kwargs))
```

This would at least reduce the number of lines :)

Reply:
I'll use a `_exec_all` function which can be called to replace all hook implementations.